lib/java/com/google/android/material/lists/item/Visual.java

```diff
@@ -0,0 +1,187 @@
+    int[] layoutMargins = calculateLayoutMargins(totalLines);
+
+    LinearLayout.LayoutParams layoutParams = (LinearLayout.LayoutParams) getLayoutParams();
+    layoutParams.gravity = gravity;
+    layoutParams.width = bounds[0];
+    layoutParams.height = bounds[1];
+    layoutParams.setMargins(layoutMargins[0], layoutMargins[1], layoutMargins[2], layoutMargins[3]);
+
+    setShapeAppearanceModel(calculateShapeAppearanceModel());
+  }
+
+  private int calculateGravity(int totalLines) {
+    int gravity = Gravity.CENTER_VERTICAL;
+    if (type == ICON || totalLines >= 3) {
+      gravity = Gravity.TOP;
+    }
+    return gravity;
+  }
+
+  private int[] calculateBounds(int totalLines) {
+    int[] bounds = new int[2];
+
+    if (type == RECTANGLE) {
+      int rectangleWidth = getDimensionInt(R.dimen.mtrl_list_item_visual_rectangle_width);
+      int rectangleHeight = getDimensionInt(R.dimen.mtrl_list_item_visual_rectangle_height);
+      bounds[0] = rectangleWidth;
+      bounds[1] = rectangleHeight;
+    } else {
+      int size = 0;
+      switch (type) {
+        case ICON:
+          size = getDimensionInt(R.dimen.mtrl_list_item_visual_icon_size);
+          break;
+        case CIRCLE:
+          size = getDimensionInt(R.dimen.mtrl_list_item_visual_circle_size);
+          break;
+        case SQUARE:
+          if (totalLines == 1) {
+            size = getDimensionInt(R.dimen.mtrl_list_item_visual_square_large_size);
+          } else {
+            size = getDimensionInt(R.dimen.mtrl_list_item_visual_square_size);
+          }
+      }
+      bounds[0] = size;
+      bounds[1] = size;
+    }
+    return bounds;
+  }
+
+  private int[] calculateLayoutMargins(int totalLines) {
+    int smallLayoutMargin = getDimensionInt(R.dimen.mtrl_list_item_visual_layout_margin_small);
+    int normalLayoutMargin = getDimensionInt(R.dimen.mtrl_list_item_visual_layout_margin_normal);
+    int largeLayoutMargin = getDimensionInt(R.dimen.mtrl_list_item_visual_layout_margin_large);
+
+    int[] layoutMargins = {
+        normalLayoutMargin,
+        normalLayoutMargin,
+        normalLayoutMargin,
+        normalLayoutMargin};
+
+    switch (type) {
+      case ICON:
+        layoutMargins[2] = largeLayoutMargin;
+        break;
+      case CIRCLE:
+      case SQUARE:
+        if (totalLines == 1) {
+          layoutMargins[1] = smallLayoutMargin;
+          layoutMargins[3] = smallLayoutMargin;
+        }
+        break;
+      case RECTANGLE:
+        layoutMargins[0] = 0;
+        if (totalLines >= 3) {
+          layoutMargins[2] = 20;
+        } else {
+          layoutMargins[1] = smallLayoutMargin;
+          layoutMargins[3] = smallLayoutMargin;
+        }
+        break;
+    }
+
+    if (ViewUtils.isLayoutRtl(this)) {
+      int temp = layoutMargins[0];
+      layoutMargins[0] = layoutMargins[2];
+      layoutMargins[2] = temp;
+    }
+    return layoutMargins;
+  }
+
+  private ShapeAppearanceModel calculateShapeAppearanceModel() {
```

Comment:
This feels a bit too inflexible and will not respond fully to theming. What happens if someone wants slightly rounded corners on a square/rectangle? (eg. The Crane example [here](https://material.io/components/lists#theming)). And what about cut corners? It feels like this needs a `<declare-styleable>` and a default style and should be using:
- `shapeAppearance` = `shapeAppearanceSmallComponent`
- `shapeAppearanceOverlay` = Custom overlay that sets corner sizes to 0dp
- If type == CIRCLE (not the biggest fan of this word as it could technically be a diamond), then use `ShapeAppearanceModel.PILL` to set corners sizes to 50% regardless of size

Reply:
I definitely agree that Material Lists should respect styling and theming. Infact, it will help shift view holder customization to XML (where it should be) and be easier for devs to understand. I will include cohesive XML style and Java APIs in the next update 🖇️